### PR TITLE
Implement move_base goals

### PR DESCRIPTION
**Pull request by tomconte**
This PR is in support of #16. /cc @ooeygui @amelhassan 

Forward/backward movement (in meters and feet) and right/left rotation (in degrees) are implemented.

I had to make two changes to the LUIS node itself:

- Modify `parseAndPublishFromJson()` to add the entities to the intent message (for rotation, the degrees are not recognised as a dimension, but they appear as a number entity).
- The keyword recogniser only worked once in my tests; after a successful keyword recognition it would just stop working. I now create a new recogniser at each iteration of the `ros::ok()` loop.

I also added the following:

- [x] Support for feet (just convert feet to meters).
- [x] Implement the STOP intent to cancel the current actions, in case the robot gets lost.

### CODE REVIEW

File: ros_msft_luis/src/azure_cs_luis.cpp [hunk 1]
```diff
@@ -120,12 +120,29 @@ void parseAndPublishFromJson(std::string luisJson)
...
+
+                                if (value_string)
+                                    value = atof(value_string);
+
                                 std::string unit = json_object_dotget_string(resolution_object, "unit");
 
                                 intent.dimension.value = value;
                                 intent.dimension.unit = unit;
                             }
+                        } else if (type == "builtin.number") {
```
Comment (ooeygui) [inline ros_msft_luis/src/azure_cs_luis.cpp:134]: We like to practice "when in rome" coding style - in this case, the brace style is next line.

File: ros_msft_luis/src/azure_cs_luis.cpp [hunk 2]
```diff
@@ -120,12 +120,29 @@ void parseAndPublishFromJson(std::string luisJson)
...
+                            JSON_Object *resolution_object = json_object_dotget_object(entity_object, "resolution");
+                            if (resolution_object)
```
Comment (ooeygui) [inline ros_msft_luis/src/azure_cs_luis.cpp:136]: Can you add a comment about what case this addresses?

File: ros_msft_luis_move_base/src/ros_msft_luis_move_base.cpp [hunk 3]
```diff
@@ -1,41 +1,67 @@
...
-const std::string LEFT = "move-left";
+const std::string FORWARD = "move forward";
+const std::string BACKWARD = "move backward";
+const std::string RIGHT = "turn right";
+const std::string LEFT = "turn left";
 
-actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction> ac("move_base", true);
+std::unique_ptr<actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction>> ac;
+
+std::string str_tolower(std::string s) {
```
Comment (ooeygui) [inline ros_msft_luis_move_base/src/ros_msft_luis_move_base.cpp:19]: brace style

**Review by ooeygui — CHANGES_REQUESTED** [review]

File: ros_msft_luis/src/azure_cs_luis.cpp (hunk 1 above)
Comment (tomconte) [reply ros_msft_luis/src/azure_cs_luis.cpp]: Done.

As a suggestion, we could use clang-format with a ROS config to ensure format consistency, e.g. run it on existing code and then using it in the CI pipeline to validate.

https://github.com/PickNikRobotics/roscpp_code_format

File: ros_msft_luis/src/azure_cs_luis.cpp (hunk 2 above)
Comment (tomconte) [reply ros_msft_luis/src/azure_cs_luis.cpp:140]: Done.

File: ros_msft_luis_move_base/src/ros_msft_luis_move_base.cpp (hunk 3 above)
Comment (tomconte) [reply ros_msft_luis_move_base/src/ros_msft_luis_move_base.cpp]: 👍🏻

**Comment by ooeygui** [comment]
Super cool :-)

**Pull request merged**